### PR TITLE
[CARBONDATA-2983][BloomDataMap] Change bloom query model to proceed multiple filter values

### PR DESCRIPTION
**Pull request by kevinjmh**
This PR is supposed to optimize the pruning for InExpression.
For a expression like colA in (1, 2, 3).
Previously it will create 3 bloom query model and for each query model it will iterate all the bloomfilters; But now we only need to generate 1 bloom query model.

Also this PR does an intersection for the pruned result generated by each expression. This optimization is important if we have multiple index columns and used them in one query.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/621/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8882/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/813/

**Comment by xuchuanyin** [comment]
This PR is supposed to optimize the pruning for InExpression.
For a expression like colA in (1, 2, 3).
Previously it will create 3 bloom query model and for each query model it will iterate all the bloomfilters; But now we only need to generate 1 bloom query model.

Also this PR does an intersection for the pruned result generated by each expression. This optimization is important if we have multiple index columns and used them in one query.

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java [hunk 1]
```diff
@@ -194,16 +186,25 @@ private String getAncestorTablePath(CarbonTable currentTable) {
...
-                  String.valueOf(bloomFilter.getBlockletNo()));
-          hitBlocklets.add(blocklet);
+        for (byte[] value: bloomQueryModel.filterValues) {
+          boolean scanRequired = bloomFilter.membershipTest(new Key(value));
+          if (scanRequired) {
+            LOGGER.debug(String.format("BloomCoarseGrainDataMap: Need to scan -> blocklet#%s",
+                String.valueOf(bloomFilter.getBlockletNo())));
+            Blocklet blocklet = new Blocklet(bloomFilter.getShardName(),
+                String.valueOf(bloomFilter.getBlockletNo()));
+            tempHitBlockletsResult.add(blocklet);
```
Comment (xuchuanyin) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java]: we can just break here if one of the values hit the blocklets -- this is a possible optimization for InExpression.
please leave a comment here if you accept the suggestion.

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java [hunk 2]
```diff
@@ -194,16 +186,25 @@ private String getAncestorTablePath(CarbonTable currentTable) {
...
+          } else {
+            LOGGER.debug(String.format("BloomCoarseGrainDataMap: Skip scan -> blocklet#%s",
```
Comment (xuchuanyin) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java]: This log is wrong here. we can only get this conclusion only if all the filter values fails the bloomfilter.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/622/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/626/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/818/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8887/

**Comment by kevinjmh** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/823/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/8892/

**Comment by xuchuanyin** [comment]
LGTM
I think this PR is OK. @ravipesala please check again

**Comment by xuchuanyin** [comment]
@ravipesala Can you check and merge this PR?

**Comment by ravipesala** [comment]
LGTM

**Pull request closed without merging**